### PR TITLE
[E219 < T1253] Map projection Cypher grammar & AST

### PR DESCRIPTION
**Pull request by antepusic**
[[Epic](https://github.com/memgraph/memgraph/pull/892) < Task] PR
- [x] Check, and update documentation if necessary
- [x] Provide the full content or a guide for the final git message
> Add map projection Cypher grammar & AST structures + tests

To keep docs changelog up to date, one more thing to do:
- [x] Write a release note here ← only the note from #892 should go in
- [ ] Tag someone from docs team in the comments

### CODE REVIEW

File: src/query/frontend/ast/cypher_main_visitor.cpp [hunk 1]
```diff
@@ -1696,6 +1696,38 @@ antlrcpp::Any CypherMainVisitor::visitMapLiteral(MemgraphCypher::MapLiteralConte
...
+  for (auto *map_el : ctx->mapElement()) {
+    if (map_el->propertyLookup()) {
+      auto key = std::any_cast<PropertyIx>(map_el->propertyLookup()->propertyKeyName()->accept(this));
+      auto property = std::any_cast<PropertyIx>(map_el->propertyLookup()->accept(this));
+      auto *property_lookup = storage_->Create<PropertyLookup>(map_projection_data.map_variable, property);
+      map_projection_data.elements.insert({key, property_lookup});
+    }
+    if (map_el->allPropertyLookup()) {
+      auto key = AddProperty("*");
+      // TODO implement AllPropertyLookup
```
Comment (antoniofilipovic) [inline src/query/frontend/ast/cypher_main_visitor.cpp:1713]: Do you plan to add implementation of allPropertyLookup in this PR?

File: src/query/frontend/ast/cypher_main_visitor.cpp [hunk 2]
```diff
@@ -2220,6 +2252,7 @@ antlrcpp::Any CypherMainVisitor::visitAtom(MemgraphCypher::AtomContext *ctx) {
     auto *where = std::any_cast<Where *>(ctx->filterExpression()->where()->accept(this));
     return static_cast<Expression *>(storage_->Create<None>(ident, list_expr, where));
   } else if (ctx->REDUCE()) {
+    // ctx->reduceExpression()->variable()
```
Comment (antoniofilipovic) [inline src/query/frontend/ast/cypher_main_visitor.cpp:2255]: Why is this commented?

File: tests/unit/cypher_main_visitor.cpp [hunk 3]
```diff
@@ -1029,6 +1029,23 @@ TEST_P(CypherMainVisitorTest, MapLiteral) {
...
+  ASSERT_TRUE(query->single_query_);
+  auto *single_query = query->single_query_;
+  auto *return_clause = dynamic_cast<Return *>(single_query->clauses_[1]);
+  auto *map_projection_literal =
+      dynamic_cast<MapProjectionLiteral *>(return_clause->body_.named_expressions[0]->expression_);
+  ASSERT_TRUE(map_projection_literal);
+  ASSERT_EQ(3, map_projection_literal->elements_.size());
+  // ast_generator.CheckLiteral(map_projection_literal->elements_[ast_generator.Prop("name")], "Morgan");
+  // ast_generator.CheckLiteral(map_projection_literal->elements_[ast_generator.Prop("lastName")], "Freeman");
+  // ast_generator.CheckLiteral(map_projection_literal->elements_[ast_generator.Prop("age")], "85");
```
Comment (antoniofilipovic) [inline tests/unit/cypher_main_visitor.cpp:1046]: There are comments here too

**Review by antoniofilipovic — COMMENTED** [review]
Looks good, but there are comments and TODOs here still

File: src/query/frontend/ast/cypher_main_visitor.cpp (hunk 1 above)
Comment (antoniofilipovic) [reply src/query/frontend/ast/cypher_main_visitor.cpp:1713]: I saw now PR for this, sorry, will close this comment

File: src/query/frontend/ast/ast.hpp [hunk 4]
```diff
@@ -1087,6 +1088,60 @@ class MapLiteral : public memgraph::query::BaseLiteral {
...
+        continue;
+      }
+
+      object->elements_[key] = entry.second->Clone(storage);
+    }
+    return object;
+  }
+
+ protected:
+  explicit MapProjectionLiteral(Expression *map_variable, const std::unordered_map<PropertyIx, Expression *> &elements)
```
Comment (antoniofilipovic) [inline src/query/frontend/ast/ast.hpp:1138]: You can put here to accept rvalue reference to `elements`, then there is no copy of `unordered_map` made here

Comment (antoniofilipovic) [inline src/query/frontend/ast/ast.hpp:1138]:
```suggestion
explicit MapProjectionLiteral(Expression *map_variable, std::unordered_map<PropertyIx, Expression *> &&elements)
      : map_variable_(map_variable), elements_(std::move(elements)) {}
```

File: src/query/frontend/ast/cypher_main_visitor.cpp (hunk 2 above)
Comment (antepusic) [reply src/query/frontend/ast/cypher_main_visitor.cpp:2255]: This is a leftover from learning how to work with the ANTLR AST, thanks for letting me know about it!

File: tests/unit/cypher_main_visitor.cpp (hunk 3 above)
Comment (antepusic) [reply tests/unit/cypher_main_visitor.cpp:1046]: Removed in #931

File: src/query/frontend/ast/cypher_main_visitor.cpp (hunk 2 above)
Comment (antepusic) [reply src/query/frontend/ast/cypher_main_visitor.cpp:2255]: Removed in #931

File: src/query/frontend/ast/ast.hpp (hunk 4 above)
Comment (antepusic) [reply src/query/frontend/ast/ast.hpp]: Fixed in #931

**Review by antoniofilipovic — APPROVED** [review]

**Pull request merged**